### DISCUSSION TITLE
DAI 2023-2024 - Java IOs benchmarking - Sergey Komarov

### DISCUSSION THREAD

**Discussion by ShadySeko** [heig-vd-dai-course/heig-vd-dai-course#57] [Show and tell]
[Click here to get to my repository](https://github.com/ShadySeko/heig-vd-dai-course-java-ios-practical-content)

Here are first of all some benchmarking results.

For Binary Files:

| File Size | Read            | Write             | Buffered Read      | Buffered Write     |
|-----------|-----------------|-------------------|----------------|----------------|
| 1B        | 0ms / 163019ns  | 0ms / 558011ns    | 0ms / 184861ns | 0ms / 575666ns |
| 1KiB      | 1ms / 1454283ns | 10ms / 10685256ns | 0ms / 438979ns | 0ms / 599494ns |
| 1MiB      | 1144ms          | 4594ms            | 28ms           | 25ms           |
| 5MiB      | 5681ms          | 22502ms           | 140ms          | 180ms          |

For Text Files:

| File Size | Read            | Write          | Buffered Read      | Buffered Write     |
|-----------|-----------------|----------------|----------------|----------------|
| 1B        | 1ms / 1259233ns | 0ms / 558008ns | 0ms / 130977ns | 0ms / 369455ns |
| 1KiB      | 1ms             | 3ms            | 0ms / 503823ns | 0ms / 743019ns |
| 1MiB      | 59ms            | 104ms          | 26ms           | 21ms           |
| 5MiB      | 168ms           | 226ms          | 131ms          | 98ms           |

&nbsp;

&nbsp;

From our benchmarking, we can see that Buffered reading and writing is better for both binary and text files, regardless of the file size.
&nbsp;

Buffered streams are more efficient, because instead of reading/writing "on the fly", a buffered implementation reads or writes only when the corresponding buffer is full (reading) or empty (writing). This minimizes the overall IO cost of the operation.

&nbsp;

The main difference between the 2 data types we have here (binary vs text) is that a binary file is nothing but a succession of bits, which don't indicate anything to us unless we can read binary on the fly (a superpower I would love to have), while a text file contains, well, **text** :)

&nbsp;

A text file contains text, but for the machine, it all starts as binary values. This is where character encoding comes in handy. The encoding format determines the way the content of a text file is encoded. Different encodings support  different amounts and types of characters.

&nbsp;

Overall, this methodology is important, as we need to know for sure how our program manages ressources. In general, we want our program to achieve the lowest time and space complexity, which in this case is determined by IO accesses. That is why we want to monitor closely how much ressources our code requires, in order to avoid unexpected behaviors like freezes or deadlocks with more complicated programs.

**Comment by ludelafo** [comment]
Great work, I have nothing more to add.
